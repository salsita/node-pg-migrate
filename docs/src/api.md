# Programmatic API

Alongside command line, you can use `node-pg-migrate` also programmatically. It exports runner function,
which takes options argument with the following structure (similar to [command line arguments](cli.md#configuration)):

## Options

> [!NOTE]
> If you use `dbClient`, you should not use `databaseUrl` at the same time and vice versa.

| Option                   | Type                                        | Description                                                                                                                                                                                                                                                                                             |
| ------------------------ | ------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `databaseUrl`            | `string or object`                          | Connection string or client config which is passed to [new pg.Client](https://node-postgres.com/api/client#constructor)                                                                                                                                                                                 |
| `dbClient`               | `pg.Client`                                 | Instance of [new pg.Client](https://node-postgres.com/api/client). Instance should be connected to DB, and after finishing migration, user is responsible to close connection                                                                                                                           |
| `migrationsTable`        | `string`                                    | The table storing which migrations have been run                                                                                                                                                                                                                                                        |
| `migrationsSchema`       | `string`                                    | The schema storing table which migrations have been run (defaults to same value as `schema`)                                                                                                                                                                                                            |
| `schema`                 | `string or array[string]`                   | The schema on which migration will be run (defaults to `public`)                                                                                                                                                                                                                                        |
| `dir`                    | `string or array[string]`                   | The directory containing your migration files. This path is resolved from `cwd()`. Alternatively, provide a [glob](https://www.npmjs.com/package/glob) pattern or an array of glob patterns and set `useGlob = true`. Note: enabling glob will read both, `dir` _and_ `ignorePattern` as glob patterns  |
| `useGlob`                | `boolean`                                   | Use [glob](https://www.npmjs.com/package/glob) to find migration files. This will use `dir` _and_ `ignorePattern` to glob-search for migration files. Note: enabling glob will read both, `dir` _and_ `ignorePattern` as glob patterns                                                                  |
| `checkOrder`             | `boolean`                                   | Check order of migrations before running them                                                                                                                                                                                                                                                           |
| `direction`              | `enum`                                      | `up` or `down`                                                                                                                                                                                                                                                                                          |
| `count`                  | `number`                                    | Amount of migration to run                                                                                                                                                                                                                                                                              |
| `timestamp`              | `boolean`                                   | Treats `count` as timestamp                                                                                                                                                                                                                                                                             |
| `ignorePattern`          | `string or array[string]`                   | Regex pattern for file names to ignore (ignores files starting with `.` by default). Alternatively, provide a [glob](https://www.npmjs.com/package/glob) pattern or an array of glob patterns and set `isGlob = true`. Note: enabling glob will read both, `dir` _and_ `ignorePattern` as glob patterns |
| `file`                   | `string`                                    | Run-only migration with this name                                                                                                                                                                                                                                                                       |
| `singleTransaction`      | `boolean`                                   | Combines all pending migrations into a single transaction so that if any migration fails, all will be rolled back (defaults to `true`)                                                                                                                                                                  |
| `createSchema`           | `boolean`                                   | Creates the configured schema if it doesn't exist                                                                                                                                                                                                                                                       |
| `createMigrationsSchema` | `boolean`                                   | Creates the configured migration schema if it doesn't exist                                                                                                                                                                                                                                             |
| `noLock`                 | `boolean`                                   | Disables locking mechanism and checks                                                                                                                                                                                                                                                                   |
| `fake`                   | `boolean`                                   | Mark migrations as run without actually performing them (use with caution!)                                                                                                                                                                                                                             |
| `dryRun`                 | `boolean`                                   |                                                                                                                                                                                                                                                                                                         |
| `log`                    | `function`                                  | Redirect log messages to this function, rather than `console`                                                                                                                                                                                                                                           |
| `logger`                 | `object with debug/info/warn/error methods` | Redirect messages to this logger object, rather than `console`                                                                                                                                                                                                                                          |
| `verbose`                | `boolean`                                   | Print all debug messages like DB queries run (if you switch it on, it will disable `logger.debug` method)                                                                                                                                                                                               |
| `decamelize`             | `boolean`                                   | Runs [`decamelize`](https://github.com/salsita/node-pg-migrate/blob/main/src/utils/decamelize.ts) on table/column/etc. names                                                                                                                                                                            |
