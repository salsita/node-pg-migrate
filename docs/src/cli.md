# CLI Usage

## Database Connection

You can specify your database connection information using [config](https://www.npmjs.com/package/config).

```jsonc
// config/default.json
{
  "db": {
    "url": "postgres://postgres:password@localhost:5432/database",
  },
}
```

or

```jsonc
// config/default.json
{
  "db": {
    "user": "postgres",
    "password": "",
    "host": "localhost",
    "port": 5432,
    "database": "database",
  },
}
```

You could also specify your database url by setting the environment variable `DATABASE_URL`.

```
DATABASE_URL=postgres://postgres@localhost/database node-pg-migrate
```

You can specify custom JSON file with config (the format is same as for `db` entry
of [config](https://www.npmjs.com/package/config) file), for example:

```jsonc
// path/to/config.json
{
  "user": "postgres",
  "password": "",
  "host": "localhost",
  "port": 5432,
  "database": "database",
}
```

> [!TIP]
> If a `.env` file exists, it will be loaded using [dotenv](https://www.npmjs.com/package/dotenv) (if installed) when
> running the node-pg-migrate binary.
> If the .env file is not on the same level where the command has been called, you
> can
> use the `--envPath` option to point to the location of your .env file.

Depending on your project's setup, it may make sense to write some custom grunt/gulp/whatever tasks that set this env
var and run your migration commands.
More on that below.

## Available Commands

| Command                                   |                                                       Description                                                       |
| ----------------------------------------- | :---------------------------------------------------------------------------------------------------------------------: |
| `node-pg-migrate create {migration-name}` | creates a new migration file with a timestamp prepended to the name you provide. Dashes replace spaces and underscores. |
| `node-pg-migrate up`                      |                                     runs all up migrations from the current state.                                      |
| `node-pg-migrate up {N}`                  |                                      runs N up migrations from the current state.                                       |
| `node-pg-migrate down`                    |                                              runs a single down migration.                                              |
| `node-pg-migrate down {N}`                |                                     runs N down migrations from the current state.                                      |
| `node-pg-migrate redo`                    |                     redoes last migration (runs a single down migration, then single up migration).                     |
| `node-pg-migrate redo {N}`                |                        redoes N last migrations (runs N down migrations, then N up migrations).                         |

## Configuration

> [!TIP]
> See all by running `node-pg-migrate --help`.
>
> Most of the configuration options can be also specified in [config](https://www.npmjs.com/package/config) file.

You can adjust defaults by passing arguments to `node-pg-migrate`:

| Argument                    | Aliases | Default                         | Description                                                                                                                                                                                                                                                                             |
| --------------------------- | ------- | ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `config-file`               | `f`     | `undefined`                     | The file with migration JSON config                                                                                                                                                                                                                                                     |
| `config-value`              |         | `db`                            | Name of config section with db options                                                                                                                                                                                                                                                  |
| `schema`                    | `s`     | `public`                        | The schema(s) on which migration will be run, used to set `search_path`                                                                                                                                                                                                                 |
| `create-schema`             |         | `false`                         | Create the configured schema if it doesn't exist                                                                                                                                                                                                                                        |
| `database-url-var`          | `d`     | `DATABASE_URL`                  | Name of env variable with database url string                                                                                                                                                                                                                                           |
| `migrations-dir`            | `m`     | `migrations`                    | The directory containing your migration files. This path is resolved from `cwd()`. Alternatively, provide a [glob](https://www.npmjs.com/package/glob) pattern and set `--use-glob`. Note: enabling glob will read both, `--migrations-dir` _and_ `--ignore-pattern` as glob patterns   |
| `use-glob`                  |         | `false`                         | Use [glob](https://www.npmjs.com/package/glob) to find migration files. This will use `--migrations-dir` _and_ `--ignore-pattern` to glob-search for migration files.                                                                                                                   |
| `migrations-schema`         |         | same value as `schema`          | The schema storing table which migrations have been run                                                                                                                                                                                                                                 |
| `create-migrations-schema`  |         | `false`                         | Create the configured migrations schema if it doesn't exist                                                                                                                                                                                                                             |
| `migrations-table`          | `t`     | `pgmigrations`                  | The table storing which migrations have been run                                                                                                                                                                                                                                        |
| `ignore-pattern`            |         | `undefined`                     | Regex pattern for file names to ignore (ignores files starting with `.` by default). Alternatively, provide a [glob](https://www.npmjs.com/package/glob) pattern and set `--use-glob`. Note: enabling glob will read both, `--migrations-dir` _and_ `--ignore-pattern` as glob patterns |
| `migration-filename-format` |         | `timestamp`                     | Choose prefix of file, `utc` (`20200605075829074`) or `timestamp` (`1591343909074`)                                                                                                                                                                                                     |
| `migration-file-language`   | `j`     | `js`                            | Language of the migration file to create (`js`, `ts` or `sql`)                                                                                                                                                                                                                          |
| `template-file-name`        |         | `undefined`                     | Utilize a custom migration template file with language inferred from its extension. The file should export the up method, accepting a MigrationBuilder instance.                                                                                                                        |
| `tsconfig`                  |         | `undefined`                     | Path to tsconfig.json. Used to setup transpiling of TS migration files. (Also sets `migration-file-language` to typescript, if not overridden)                                                                                                                                          |
| `envPath`                   |         | `same level where it's invoked` | Retrieve the path to a .env file. This feature proves handy when dealing with nested projects or when referencing a global .env file.                                                                                                                                                   |
| `timestamp`                 |         | `false`                         | Treats number argument to up/down migration as timestamp (running up migrations less or equal to timestamp or down migrations greater or equal to timestamp)                                                                                                                            |
| `check-order`               |         | `true`                          | Check order of migrations before running them, to switch it off supply `--no-check-order`                                                                                                                                                                                               |
| `single-transaction`        |         | `true`                          | Combines all pending migrations into a single transaction so that if any migration fails, all will be rolled back, to switch it off supply `--no-single-transaction`                                                                                                                    |
| `no-lock`                   |         | `false`                         | Disables locking mechanism and checks                                                                                                                                                                                                                                                   |
| `fake`                      |         | `false`                         | Mark migrations as run without actually performing them, (use with caution!)                                                                                                                                                                                                            |
| `decamelize`                |         | `false`                         | Runs `decamelize` on table/column/etc. names                                                                                                                                                                                                                                            |
| `verbose`                   |         | `true`                          | Print all debug messages like DB queries run, to switch it off supply `--no-verbose`                                                                                                                                                                                                    |
| `reject-unauthorized`       |         | `undefined`                     | Sets ssl `rejectUnauthorized` parameter. Use for e.g. self-signed certificates on the server. [see](https://node-postgres.com/announcements#2020-02-25)                                                                                                                                 |

For SSL connection to DB you can set `PGSSLMODE` environment variable to value
from [list](https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNECT-SSLMODE) other
than `disable`.
e.g. `PGSSLMODE=require node-pg-migrate up` ([pg](https://github.com/brianc/node-postgres/blob/main/CHANGELOG.md#v260)
will take it into account)

### JSON Configuration

> [!TIP]
> You can use [config](https://www.npmjs.com/package/config) or your own json file with configuration
> (`config-file` command line option).

You can also specify your database connection here, [see](#database-connection).

Other available options are:

```jsonc
{
  "schema": "public",
  "createSchema": false,
  "migrationsDir": "migrations",
  "migrationsSchema": "public",
  "createMigrationsSchema": false,
  "migrationsTable": "pgmigrations",
  "migrationFilenameFormat": "utc",
  "migrationFileLanguage": "js",
  "ignorePattern": undefined,
  "checkOrder": true,
  "verbose": true,
  "decamelize": false,
  "tsconfig": "tsconfig.json",
}
```
