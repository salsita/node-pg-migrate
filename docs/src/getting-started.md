# Getting Started

## Preconditions

- Node.js 20.11 or higher
- PostgreSQL 13 or higher (lower versions may work but are not supported officially)

If you don't already have the [`pg`](https://node-postgres.com/) library installed, you will need to add pg as either a direct or dev dependency

::: code-group

```sh [npm]
$ npm add pg
```

```sh [pnpm]
$ pnpm add pg
```

```sh [yarn]
$ yarn add pg
```

```sh [bun]
$ bun add pg
```

:::

## Installation

::: code-group

```sh [npm]
$ npm add -D node-pg-migrate
```

```sh [pnpm]
$ pnpm add -D node-pg-migrate
```

```sh [yarn]
$ yarn add -D node-pg-migrate
```

```sh [bun]
$ bun add -D node-pg-migrate
```

:::

Installing this module adds a runnable file into your `node_modules/.bin` directory. If installed globally (with the -g option), you can run `node-pg-migrate` and if not, you can run `./node_modules/.bin/node-pg-migrate.js`

## Quick Example

> [!IMPORTANT]
> This example assumes you are using `npm` as your package manager. If you are using another package manager, replace `npm` with the appropriate command.

Add `node-pg-migrate` to `scripts` section of your `package.json` so you are able to quickly run commands.

```jsonc
{
  "scripts": {
    // ..
    "migrate": "node-pg-migrate", // [!code ++]
  },
}
```

Run `npm run migrate create my-first-migration`. It will create file `xxx_my-first-migration.js` in `migrations` folder.  
Open it and change contents to:

```js
export const up = (pgm) => {
  pgm.createTable('users', {
    id: 'id',
    name: { type: 'varchar(1000)', notNull: true },
    createdAt: {
      type: 'timestamp',
      notNull: true,
      default: pgm.func('current_timestamp'),
    },
  });
  pgm.createTable('posts', {
    id: 'id',
    userId: {
      type: 'integer',
      notNull: true,
      references: '"users"',
      onDelete: 'cascade',
    },
    body: { type: 'text', notNull: true },
    createdAt: {
      type: 'timestamp',
      notNull: true,
      default: pgm.func('current_timestamp'),
    },
  });
  pgm.createIndex('posts', 'userId');
};
```

Save migration file.

Now you should put your DB connection string to `DATABASE_URL` environment variable and run `npm run migrate up`.
(e.g. `DATABASE_URL=postgres://test:test@localhost:5432/test npm run migrate up`)

You should now have two tables in your DB :tada:

If you want to change your schema later, you can e.g. add lead paragraph to posts:

Run `npm run migrate create posts_lead`, edit `xxx_posts_lead.js`:

```js
export const up = (pgm) => {
  pgm.addColumns('posts', {
    lead: { type: 'text', notNull: true },
  });
};
```

Run `npm run migrate up` and there will be a new column in `posts` table :tada:

## Want to know more?

- [CLI commands](/cli)
- [Programmatic API](/api)
- [Migration files](/migrations/)
