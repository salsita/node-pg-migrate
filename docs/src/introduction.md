# What is node-pg-migrate?

Node.js database migration management built exclusively for postgres. (But can also be used for other DBs conforming to SQL standard - e.g. [CockroachDB](https://github.com/cockroachdb/cockroach).)  
Started by [Theo Ephraim](https://github.com/theoephraim/), then handed over to [Salsita Software](https://www.salsitasoft.com/) and now maintained by [@Shinigami92](https://github.com/Shinigami92).

## Explanation & Goals

### _Why only Postgres?_

By writing this migration tool specifically for postgres instead of accommodating many databases, we can actually provide a full featured tool that is much simpler to use and maintain. I was tired of using crippled database tools just in case one day we switch our database.

### _Async / Sync_

Everything is async in node, and that's great, but a migration tool should really just be a fancy wrapper that generates SQL. Most other migration tools force you to bring in control flow libraries or wrap everything in callbacks as soon as you want to do more than a single operation in a migration. Plus by building up a stack of operations, we can automatically infer down migrations (sometimes) to save even more time.

### _Naming / Raw Sql_

Many tools force you to use their constants to do things like specify data types. Again, this tool should be a fancy wrapper that generates SQL, so whenever possible, it should just pass through user values directly to the SQL. The hard part is remembering the syntax of the specific operation, not remembering how to type "timestamp"!

## License

This project is licensed under [MIT](https://github.com/salsita/node-pg-migrate/blob/main/LICENSE) license.
