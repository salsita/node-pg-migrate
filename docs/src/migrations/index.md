---
next:
  text: 'Tables'
  link: '/migrations/tables'
---

# Defining Migrations

## Running Migrations

When you run `node-pg-migrate create` a new migration file is created that looks like this:

```javascript
export const shorthands = undefined;

export const up = function up(pgm) {};

export const down = function down(pgm) {};
```

`pgm` is a helper object that provides migration operations and `run` is the callback to call when you are done.

`shorthands` is optional for column type shorthands. You can specify custom types which will be expanded to column
definition.

### Example

```js
export const shorthands = {
  id: { type: 'uuid', primaryKey: true },
  createdAt: {
    type: 'timestamp',
    notNull: true,
    default: new PgLiteral('current_timestamp'),
  },
};
```

It will in `pgm.createTable('test', { id: 'id', createdAt: 'createdAt' });` produce SQL

```sql
CREATE TABLE "test" ("id" uuid PRIMARY KEY, "createdAt" timestamp DEFAULT current_timestamp NOT NULL);
```

These shorthands are inherited from previous migrations. You can override/change value by simply defining a new value
for a given shorthand name
if it is used in current and all following migrations (until changed again).

> [!IMPORTANT]
> Calling the migration functions on `pgm` doesn't migrate your database. These functions just add sql commands to a
> stack that is run.

## Automatic Down Migrations

If `export const down` is not present in a migration, node-pg-migrate will try to automatically infer the operations that
make up the down migration by reversing the operations of the up migration. Only some operations have automatically
inferrable equivalents (the details below on each operation). Sometimes, migrations are destructive and cannot be rolled
back. In this case, you can set `export const down = false` to tell node-pg-migrate that the down migration is impossible.

## Async Migrations

In some cases, you may want to perform some async operation during a migration, for example, fetching some information
from an external server, or inserting some data into the database. To make a migration block operate in async mode, add
another callback argument to the function signature. However, be aware that NONE of the pgm operations will be executed
until `run()` is called. Here's an example:

```javascript
export const up = function up(pgm, run) {
  doSomethingAsync(function () {
    run();
  });
};
```

Another way how to perform some async operation is to return [Promise](https://promisesaplus.com/) from `up` or `down`
function. Example:

```javascript
export const up = function (pgm) {
  return new Promise((resolve) => {
    // doSomethingAsync
    resolve();
  });
};
```

or

```javascript
export const up = async (pgm) => {
  // doSomethingAsync
};
```

## Using schemas

Instead of passing string as name to `pgm` functions, you can pass an object with keys `schema` and `name`. E.g.

```javascript
pgm.createTable(
  { schema: 'my_schema', name: 'my_table_name' },
  { id: 'serial' }
);
```

will generate

```sql
CREATE TABLE "my_schema"."my_table_name" (
 "id" serial
);
```

### Type

```ts
type Name = string | { schema: string; name: string };
```

## Locking

`node-pg-migrate` automatically checks if no other migration is running. To do so, it uses an
[advisory lock](https://www.postgresql.org/docs/current/static/explicit-locking.html#id-1.5.12.6.9.2)
(see [#239](https://github.com/salsita/node-pg-migrate/pull/239)).
Lock is held for the duration of DB session, so if migration scripts froze up, you need to kill it,
before running another migration script.

## Migration methods

The `pgm` object that is passed to each up/down block has many different operations available.
Each operation is simply
a function that generates some sql and stores it in the current pgm context.

By default, all migrations will be run in a transaction.
To disable transactions for a specific migration,
call `pgm.noTransaction()`
This is required for some SQL operations that cannot be run within a transaction.
It should be used carefully.

- [Tables](tables.md)
- [Columns](columns.md)
- [Constraints](constraints.md)
- [Indexes](indexes.md)
- [Functions](functions.md)
- [Triggers](triggers.md)
- [Schemas](schemas.md)
- [Sequences](sequences.md)
- [Views](views.md)
- [Materialized Views](mViews.md)
- [Types](types.md)
- [Domains](domains.md)
- [Operators](operators.md)
- [Roles](roles.md)
- [Policies](policies.md)
- [Extensions](extensions.md)
- [Miscellaneous](misc.md)
