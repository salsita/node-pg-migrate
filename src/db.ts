/*
 This file just manages the database connection and provides a query method
 */
import { inspect } from 'node:util';
import type {
  Client,
  ClientBase,
  ClientConfig,
  QueryArrayConfig,
  QueryArrayResult,
  QueryConfig,
  QueryResult,
} from 'pg';
// This needs to be imported as `*`, otherwise it will fail in esm
import * as pg from 'pg';
import type { DB, Logger } from './types';

/* eslint-disable @typescript-eslint/no-explicit-any */
export interface DBConnection extends DB {
  createConnection(): Promise<void>;

  column(
    columnName: string,
    queryConfig: QueryArrayConfig,
    values?: any[]
  ): Promise<any[]>;
  column(columnName: string, queryConfig: QueryConfig): Promise<any[]>;
  column(
    columnName: string,
    queryTextOrConfig: string | QueryConfig,
    values?: any[]
  ): Promise<any[]>;

  connected: () => boolean;

  addBeforeCloseListener: (listener: any) => number;

  close(): Promise<void>;
}

enum ConnectionStatus {
  DISCONNECTED = 'DISCONNECTED',
  CONNECTED = 'CONNECTED',
  ERROR = 'ERROR',
}

function db(
  connection: ClientBase | string | ClientConfig,
  logger: Logger = console
): DBConnection {
  const isExternalClient =
    typeof connection === 'object' &&
    'query' in connection &&
    typeof connection.query === 'function';
  let connectionStatus = ConnectionStatus.DISCONNECTED;

  const client: Client = isExternalClient
    ? (connection as Client)
    : new pg.Client(connection as string | ClientConfig);

  const beforeCloseListeners: any[] = [];

  const createConnection: DBConnection['createConnection'] = () =>
    new Promise((resolve, reject) => {
      if (isExternalClient || connectionStatus === ConnectionStatus.CONNECTED) {
        resolve();
      } else if (connectionStatus === ConnectionStatus.ERROR) {
        reject(
          new Error('Connection already failed, do not try to connect again')
        );
      } else {
        client.connect((err) => {
          if (err) {
            connectionStatus = ConnectionStatus.ERROR;
            logger.error(`could not connect to postgres: ${inspect(err)}`);
            reject(err);
            return;
          }

          connectionStatus = ConnectionStatus.CONNECTED;
          resolve();
        });
      }
    });

  const query: DBConnection['query'] = async (
    queryTextOrConfig: string | QueryConfig | QueryArrayConfig,
    values?: any[]
  ): Promise<QueryArrayResult | QueryResult> => {
    await createConnection();
    try {
      return await client.query(queryTextOrConfig, values);
    } catch (error: any) {
      const { message, position }: { message: string; position: number } =
        error;

      const string: string =
        typeof queryTextOrConfig === 'string'
          ? queryTextOrConfig
          : queryTextOrConfig.text;

      if (message && position >= 1) {
        const endLineWrapIndexOf = string.indexOf('\n', position);
        const endLineWrapPos =
          endLineWrapIndexOf >= 0 ? endLineWrapIndexOf : string.length;
        const stringStart = string.slice(0, endLineWrapPos);
        const stringEnd = string.slice(endLineWrapPos);
        const startLineWrapPos = stringStart.lastIndexOf('\n') + 1;
        const padding = ' '.repeat(position - startLineWrapPos - 1);
        logger.error(`Error executing:
${stringStart}
${padding}^^^^${stringEnd}

${message}
`);
      } else {
        logger.error(`Error executing:
${string}
${error}
`);
      }

      throw error;
    }
  };

  const select: DBConnection['select'] = async (
    queryTextOrConfig: string | QueryConfig | QueryArrayConfig,
    values?: any[]
  ) => {
    const { rows } = await query(queryTextOrConfig, values);
    return rows;
  };

  const column: DBConnection['column'] = async (
    columnName: string,
    queryTextOrConfig: string | QueryConfig | QueryArrayConfig,
    values?: any[]
  ) => {
    const rows = await select(queryTextOrConfig, values);
    return rows.map((r: { [key: string]: any }) => r[columnName]);
  };

  return {
    createConnection,
    query,
    select,
    column,

    connected: () => connectionStatus === ConnectionStatus.CONNECTED,
    addBeforeCloseListener: (listener) => beforeCloseListeners.push(listener),
    close: async () => {
      await beforeCloseListeners.reduce(
        (promise, listener) =>
          promise.then(listener).catch((error: any) => {
            logger.error(error.stack || error);
          }),
        Promise.resolve()
      );
      if (!isExternalClient) {
        connectionStatus = ConnectionStatus.DISCONNECTED;
        client.end();
      }
    },
  };
}

export default db;
