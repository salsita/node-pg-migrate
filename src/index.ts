export { Migration } from './migration';
export { MigrationBuilder } from './migrationBuilder';
export type {
  CreateCast,
  CreateCastFn,
  CreateCastOptions,
  CreateCastWithFunctionOptions,
  CreateCastWithInoutOptions,
  CreateCastWithoutFunctionOptions,
  DropCast,
  DropCastOptions,
} from './operations/casts';
export type {
  AlterDomain,
  AlterDomainOptions,
  CreateDomain,
  CreateDomainFn,
  CreateDomainOptions,
  DomainOptions,
  DropDomain,
  DropDomainOptions,
  RenameDomain,
  RenameDomainFn,
} from './operations/domains';
export type {
  CreateExtension,
  CreateExtensionFn,
  CreateExtensionOptions,
  DropExtension,
  DropExtensionOptions,
  Extension,
  StringExtension,
} from './operations/extensions';
export type {
  CreateFunction,
  CreateFunctionFn,
  CreateFunctionOptions,
  DropFunction,
  DropFunctionOptions,
  FunctionOptions,
  FunctionParam,
  FunctionParamType,
  RenameFunction,
  RenameFunctionFn,
} from './operations/functions';
export type {
  CascadeOption,
  DropOptions,
  IfExistsOption,
  IfNotExistsOption,
  LiteralUnion,
  Name,
  Nullable,
  Operation,
  OperationFn,
  PublicPart,
  Reversible,
  Type,
  Value,
} from './operations/generalTypes';
export type {
  AllTablesOptions,
  CommonGrantOnTablesOptions,
  CommonOnTablesOptions,
  GrantOnAllTablesOptions,
  GrantOnSchemas,
  GrantOnSchemasFn,
  GrantOnSchemasOptions,
  GrantOnSomeTablesOptions,
  GrantOnTables,
  GrantOnTablesFn,
  GrantOnTablesOptions,
  GrantRoles,
  GrantRolesFn,
  GrantRolesOptions,
  OnlyAdminOption,
  OnlyGrantOnSchemasOptions,
  OnlyGrantOption,
  RevokeOnObjectsOptions,
  RevokeOnSchemas,
  RevokeOnSchemasOptions,
  RevokeOnTables,
  RevokeOnTablesOptions,
  RevokeRoles,
  RevokeRolesOptions,
  SchemaPrivilege,
  SomeTablesOptions,
  TablePrivilege,
  WithAdminOption,
  WithGrantOption,
} from './operations/grants';
export type {
  CreateIndex,
  CreateIndexFn,
  CreateIndexOptions,
  DropIndex,
  DropIndexOptions,
  IndexColumn,
} from './operations/indexes';
export type {
  AlterMaterializedView,
  AlterMaterializedViewOptions,
  CreateMaterializedView,
  CreateMaterializedViewFn,
  CreateMaterializedViewOptions,
  DropMaterializedView,
  DropMaterializedViewOptions,
  RefreshMaterializedView,
  RefreshMaterializedViewFn,
  RefreshMaterializedViewOptions,
  RenameMaterializedView,
  RenameMaterializedViewColumn,
  RenameMaterializedViewColumnFn,
  RenameMaterializedViewFn,
  StorageParameters,
} from './operations/materializedViews';
export type {
  AddToOperatorFamily,
  AddToOperatorFamilyFn,
  CreateOperator,
  CreateOperatorClass,
  CreateOperatorClassFn,
  CreateOperatorClassOptions,
  CreateOperatorFamily,
  CreateOperatorFamilyFn,
  CreateOperatorFamilyOptions,
  CreateOperatorFn,
  CreateOperatorOptions,
  DropOperator,
  DropOperatorClass,
  DropOperatorClassOptions,
  DropOperatorFamily,
  DropOperatorFamilyOptions,
  DropOperatorOptions,
  OperatorListDefinition,
  RemoveFromOperatorFamily,
  RenameOperatorClass,
  RenameOperatorClassFn,
  RenameOperatorFamily,
  RenameOperatorFamilyFn,
} from './operations/operators';
export type {
  AlterPolicy,
  CreatePolicy,
  CreatePolicyOptions,
  CreatePolicyOptionsEn,
  DropPolicy,
  DropPolicyOptions,
  PolicyOptions,
  RenamePolicy,
  RenamePolicyFn,
} from './operations/policies';
export type {
  AlterRole,
  CreateRole,
  CreateRoleFn,
  CreateRoleOptions,
  DropRole,
  DropRoleOptions,
  RenameRole,
  RenameRoleFn,
  RoleOptions,
} from './operations/roles';
export type {
  CreateSchema,
  CreateSchemaFn,
  CreateSchemaOptions,
  DropSchema,
  DropSchemaOptions,
  RenameSchema,
  RenameSchemaFn,
} from './operations/schemas';
export type {
  AlterSequence,
  AlterSequenceOptions,
  CreateSequence,
  CreateSequenceFn,
  CreateSequenceOptions,
  DropSequence,
  DropSequenceOptions,
  RenameSequence,
  RenameSequenceFn,
  SequenceOptions,
} from './operations/sequences';
export type { Sql } from './operations/sql';
export type {
  Action,
  AddColumns,
  AddColumnsFn,
  AddColumnsOptions,
  AlterColumn,
  AlterColumnOptions,
  AlterTable,
  AlterTableOptions,
  ColumnDefinition,
  ColumnDefinitions,
  ConstraintOptions,
  CreateConstraint,
  CreateConstraintFn,
  CreateTable,
  CreateTableFn,
  DropColumns,
  DropColumnsOptions,
  DropConstraint,
  DropConstraintOptions,
  DropTable,
  DropTableOptions,
  ForeignKeyOptions,
  Like,
  LikeOptions,
  ReferencesOptions,
  RenameColumn,
  RenameColumnFn,
  RenameConstraint,
  RenameConstraintFn,
  RenameTable,
  RenameTableFn,
  SequenceGeneratedOptions,
  TableOptions,
} from './operations/tables';
export type {
  CreateTrigger,
  CreateTriggerFn,
  CreateTriggerFn1,
  CreateTriggerFn2,
  DropTrigger,
  DropTriggerOptions,
  RenameTrigger,
  RenameTriggerFn,
  TriggerOptions,
} from './operations/triggers';
export type {
  AddTypeAttribute,
  AddTypeAttributeFn,
  AddTypeValue,
  AddTypeValueOptions,
  CreateType,
  CreateTypeFn,
  DropType,
  DropTypeAttribute,
  DropTypeAttributeOptions,
  DropTypeOptions,
  RenameType,
  RenameTypeAttribute,
  RenameTypeAttributeFn,
  RenameTypeFn,
  RenameTypeValue,
  RenameTypeValueFn,
  SetTypeAttribute,
} from './operations/types';
export type {
  AlterView,
  AlterViewColumn,
  AlterViewColumnOptions,
  AlterViewOptions,
  CreateView,
  CreateViewFn,
  CreateViewOptions,
  DropView,
  DropViewOptions,
  RenameView,
  RenameViewFn,
  ViewOptions,
} from './operations/views';
export { PgType } from './pgType';
export { runner } from './runner';
export type { RunnerOption } from './runner';
export { PgLiteral, isPgLiteral } from './utils';
export type { PgLiteralValue } from './utils';
