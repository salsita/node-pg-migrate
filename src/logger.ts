export type LogFn = (msg: string) => void;

export type Logger = {
  debug?: LogFn;
  info: LogFn;
  warn: LogFn;
  error: LogFn;
};
