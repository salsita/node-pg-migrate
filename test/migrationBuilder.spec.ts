import { describe, expect, it, vi } from 'vitest';
import { MigrationBuilder } from '../src';

describe('migrationBuilder', () => {
  it('should expose MigrationBuilder to allow using as sql builder', () => {
    const pgm = new MigrationBuilder(
      {
        query: vi.fn(),
        select: vi.fn(),
      },
      undefined,
      true,
      console
    );

    pgm.createTable('users', {
      id: 'id',
      name: { type: 'varchar(1000)', notNull: true },
      createdAt: {
        type: 'timestamp',
        notNull: true,
        default: pgm.func('current_timestamp'),
      },
    });
    pgm.createTable('posts', {
      id: 'id',
      userId: {
        type: 'integer',
        notNull: true,
        references: '"users"',
        onDelete: 'CASCADE',
      },
      body: { type: 'text', notNull: true },
      createdAt: {
        type: 'timestamp',
        notNull: true,
        default: pgm.func('current_timestamp'),
      },
    });
    pgm.createIndex('posts', 'userId');

    expect(pgm.getSql()).toMatchSnapshot();
  });
});
