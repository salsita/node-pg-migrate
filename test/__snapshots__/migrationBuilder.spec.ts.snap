// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrationBuilder > should expose MigrationBuilder to allow using as sql builder 1`] = `
"CREATE TABLE "users" (
  "id" serial PRIMARY KEY,
  "name" varchar(1000) NOT NULL,
  "created_at" timestamp DEFAULT current_timestamp NOT NULL
);
CREATE TABLE "posts" (
  "id" serial PRIMARY KEY,
  "user_id" integer NOT NULL REFERENCES "users" ON DELETE CASCADE,
  "body" text NOT NULL,
  "created_at" timestamp DEFAULT current_timestamp NOT NULL
);
CREATE INDEX "posts_user_id_index" ON "posts" ("user_id");
"
`;
